backend/src/apiserver/resource/resource_manager.go

```diff
@@ -195,616 +510,1147 @@ func (r *ResourceManager) ArchiveExperiment(ctx context.Context, experimentId st
-		return util.NewInternalServerError(err, "Failed to decompress workflow")
+		return util.NewInternalServerError(err, "Failed to retry run %s due to error decompressing execution spec", runId)
 	}
 
 	if err := execSpec.CanRetry(); err != nil {
-		return err
+		return util.NewInternalServerError(err, "Failed to retry run %s as it does not allow reties", runId)
 	}
 
 	newExecSpec, podsToDelete, err := execSpec.GenerateRetryExecution()
 	if err != nil {
-		return util.Wrap(err, "Retry run failed.")
+		return util.Wrapf(err, "Failed to retry run %s", runId)
 	}
 
 	if err = deletePods(ctx, r.k8sCoreClient, podsToDelete, namespace); err != nil {
-		return util.NewInternalServerError(err, "Retry run failed. Failed to clean up the failed pods from previous run.")
+		return util.NewInternalServerError(err, "Failed to retry run %s due to error cleaning up the failed pods from the previous attempt", runId)
 	}
 
 	// First try to update workflow
-	updateError := r.updateWorkflow(ctx, newExecSpec, namespace)
+	// If fail to get the workflow, return error.
+	latestWorkflow, updateError := r.getWorkflowClient(namespace).Get(ctx, newExecSpec.ExecutionName(), v1.GetOptions{})
+	if updateError == nil {
+		// Update the workflow's resource version to latest.
+		newExecSpec.SetVersion(latestWorkflow.Version())
+		_, updateError = r.getWorkflowClient(namespace).Update(ctx, newExecSpec, v1.UpdateOptions{})
+	}
 	if updateError != nil {
 		// Remove resource version
 		newExecSpec.SetVersion("")
 		newCreatedWorkflow, createError := r.getWorkflowClient(namespace).Create(ctx, newExecSpec, v1.CreateOptions{})
 		if createError != nil {
-			return util.NewInternalServerError(createError,
-				"Retry run failed. Failed to create or update the run. Update Error: %s, Create Error: %s",
-				updateError.Error(), createError.Error())
+			return util.Wrap(
+				util.NewInternalServerError(updateError, "Failed to retry run %s due to error updating the old workflow", runId),
+				util.NewInternalServerError(createError, "Failed to retry run %s due to error creating a new workflow", runId).Error(),
+			)
 		}
 		newExecSpec = newCreatedWorkflow
 	}
-	err = r.runStore.UpdateRun(runId, string(newExecSpec.ExecutionStatus().Condition()), 0, newExecSpec.ToStringForStore())
+	condition := string(newExecSpec.ExecutionStatus().Condition())
+	state := model.RuntimeState(condition).ToV2().ToString()
+	err = r.runStore.UpdateRun(runId, condition, 0, newExecSpec.ToStringForStore(), state)
 	if err != nil {
-		return util.NewInternalServerError(err, "Failed to update the database entry.")
+		return util.NewInternalServerError(err, "Failed to retry run %s due to error updating entry", runId)
 	}
 	return nil
 }
 
-func (r *ResourceManager) ReadLog(ctx context.Context, runId string, nodeId string, follow bool, dst io.Writer) error {
-	run, err := r.checkRunExist(runId)
+// Deletes a run entry with a given id.
+func (r *ResourceManager) DeleteRun(ctx context.Context, runId string) error {
```

Comment:
It looks like many diffs are due to that we're moving functions around, could we keep them in the original order, so that we see less diffs here? (It's really hard to tell if the functions being moved has any change from the original)

Reply:
That's a good point. I have it arranged in CRUD, grouped by resource type, in a hierarchical order. I guess the file is too old to have such refactoring. Let me change it back later today.